lib/utils.pm

```diff
@@ -463,6 +468,102 @@ sub pkcon_quit {
     script_run("pkcon quit; while pgrep packagekitd; do sleep 1; done");
 }
 
+sub register_needle_tags {
+    my ($tag) = @_;
+    my @a = @{needle::tags($tag)};
+    for my $n (@a) { $n->register(); }
+}
+
+sub unregister_needle_tags {
+    my ($tag) = @_;
+    my @a = @{needle::tags($tag)};
+    for my $n (@a) { $n->unregister(); }
+}
+
+sub remove_desktop_needles {
+    my $desktop = shift;
+    if (!check_var("DESKTOP", $desktop) && !check_var("FULL_DESKTOP", $desktop)) {
+        unregister_needle_tags("ENV-DESKTOP-$desktop");
+    }
+}
+
+sub is_sle_server {
+    return is_sles4sap() || get_var('FLAVOR', '') =~ /^Server/;
+}
+
+sub is_sle_desktop {
+    return get_var('FLAVOR', '') =~ /^Desktop/;
+}
+
+sub is_sles4sap {
+    return get_var('FLAVOR', '') =~ /SAP/;
+}
+
+sub cleanup_sle_needles {
+    remove_desktop_needles("lxde");
+    remove_desktop_needles("kde");
+    remove_desktop_needles("gnome");
+    remove_desktop_needles("xfce");
+    remove_desktop_needles("minimalx");
+    remove_desktop_needles("textmode");
+
+    if (!check_var("VIDEOMODE", "text")) {
+        unregister_needle_tags("ENV-VIDEOMODE-text");
+    }
+
+    if (get_var("INSTLANG") && get_var("INSTLANG") ne "en_US") {
+        unregister_needle_tags("ENV-INSTLANG-en_US");
+    }
+    else {    # english default
+        unregister_needle_tags("ENV-INSTLANG-de_DE");
+    }
+
+    if (get_var('VERSION', '') ne '12') {
+        unregister_needle_tags("ENV-VERSION-12");
+    }
+
+    if (get_var('VERSION', '') ne '12-SP1') {
+        unregister_needle_tags("ENV-VERSION-12-SP1");
+    }
+
+    if (get_var('VERSION', '') ne '12-SP2') {
+        unregister_needle_tags("ENV-VERSION-12-SP2");
+    }
+
+    my $tounregister = sle_version_at_least('12-SP2') ? '0' : '1';
+    unregister_needle_tags("ENV-SP2ORLATER-$tounregister");
+
+    if (!is_sle_server) {
+        unregister_needle_tags("ENV-FLAVOR-Server-DVD");
+    }
+
+    if (!is_sle_desktop) {
+        unregister_needle_tags("ENV-FLAVOR-Desktop-DVD");
+    }
+
+    if (!is_jeos) {
+        unregister_needle_tags('ENV-FLAVOR-JeOS-for-kvm');
+    }
+    if (!check_var("ARCH", "s390x")) {
+        unregister_needle_tags('ENV-ARCH-s390x');
+    }
+
+    if (get_var('OFW')) {
+        unregister_needle_tags('ENV-OFW-0');
+    }
+    else {
+        unregister_needle_tags('ENV-OFW-1');
+    }
+}
+
+sub reload_all_needles {
+    for my $n (needle->all()) {
+        $n->unregister();
+    }
+    needle::init(get_required_var('PRODUCTDIR') . '/needles');
```

Comment:
this code runs in the test thread - but the backend is actually the one checking them. So I wonder how this can have worked - if it did